pkg/resources/function.go

```diff
@@ -161,402 +168,663 @@ func Function() *schema.Resource {
 			}
 		} else {
-			q, err := builder.Unsecure()
-			if err != nil {
-				return err
-			}
-			if err = snowflake.Exec(db, q); err != nil {
-				return fmt.Errorf("error unsetting secure for function %v", d.Id())
+			if err := client.Functions.Alter(ctx, sdk.NewAlterFunctionRequest(id.name, id.argumentTypes).WithUnsetSecure(sdk.Bool(true))); err != nil {
+				return diag.FromErr(err)
 			}
 		}
 	}
-
 	if d.HasChange("comment") {
 		comment := d.Get("comment")
-
-		if c := comment.(string); c == "" {
-			q, err := builder.RemoveComment()
-			if err != nil {
-				return err
-			}
-			if err := snowflake.Exec(db, q); err != nil {
-				return fmt.Errorf("error unsetting comment for function %v err = %w", d.Id(), err)
+		if comment != "" {
+			if err := client.Functions.Alter(ctx, sdk.NewAlterFunctionRequest(id.name, id.argumentTypes).WithSetComment(sdk.String(comment.(string)))); err != nil {
+				return diag.FromErr(err)
 			}
 		} else {
-			q, err := builder.ChangeComment(c)
-			if err != nil {
-				return err
-			}
-			if err := snowflake.Exec(db, q); err != nil {
-				return fmt.Errorf("error updating comment for function %v err = %w", d.Id(), err)
+			if err := client.Functions.Alter(ctx, sdk.NewAlterFunctionRequest(id.name, id.argumentTypes).WithUnsetComment(sdk.Bool(true))); err != nil {
+				return diag.FromErr(err)
 			}
 		}
 	}
-
-	return ReadFunction(d, meta)
+	return ReadContextFunction(ctx, d, meta)
 }
 
-// DeleteFunction implements schema.DeleteFunc.
-func DeleteFunction(d *schema.ResourceData, meta interface{}) error {
+func DeleteContextFunction(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	db := meta.(*sql.DB)
-	pID, err := splitFunctionID(d.Id())
-	if err != nil {
-		return err
-	}
-	builder := snowflake.NewFunctionBuilder(
-		pID.DatabaseName,
-		pID.SchemaName,
-		pID.FunctionName,
-		pID.ArgTypes,
-	)
+	client := sdk.NewClientFromDB(db)
 
-	q, err := builder.Drop()
+	id, err := DecodeFunctionID(d.Id())
 	if err != nil {
-		return err
+		return diag.FromErr(err)
 	}
-	if err := snowflake.Exec(db, q); err != nil {
-		return fmt.Errorf("error deleting function %v err = %w", d.Id(), err)
+	if err := client.Functions.Drop(ctx, sdk.NewDropFunctionRequest(id.name, id.argumentTypes)); err != nil {
+		return diag.FromErr(err)
 	}
-
 	d.SetId("")
-
 	return nil
 }
 
+func parseFunctionArguments(d *schema.ResourceData) ([]sdk.FunctionArgumentRequest, diag.Diagnostics) {
+	args := make([]sdk.FunctionArgumentRequest, 0)
+	if v, ok := d.GetOk("arguments"); ok {
+		for _, arg := range v.([]interface{}) {
+			argName := arg.(map[string]interface{})["name"].(string)
+			argType := arg.(map[string]interface{})["type"].(string)
+			argDataType, diags := convertFunctionDataType(argType)
+			if diags != nil {
+				return nil, diags
+			}
+			args = append(args, sdk.FunctionArgumentRequest{ArgName: argName, ArgDataType: argDataType})
+		}
+	}
+	return args, nil
+}
+
 type functionID struct {
```

Comment:
Is there a point in keeping this ID? we have helpers for encoding and decoding to replace such IDs. At first sight, there's no special functionality, can we replace them with our helper functions?

Reply:
no, it was waiting on a fix from proc with identifier helpers, but I just added here to this PR

Reply:
Oh, ok 👍